pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -243,20 +243,24 @@ func (t *Table) getOrCreateIndexSet(id string, async bool) (IndexSet, error) {
 	t.indexSetsMtx.Lock()
 
 	indexSet, ok = t.indexSets[id]
-	if !ok {
-		var err error
-		baseIndexSet := t.baseUserIndexSet
-		if id == "" {
-			baseIndexSet = t.baseCommonIndexSet
-		}
+	if ok {
+		t.indexSetsMtx.Unlock()
+		return indexSet, nil
+	}
 
-		indexSet, err = NewIndexSet(t.name, id, filepath.Join(t.cacheLocation, id), baseIndexSet, t.boltDBIndexClient, t.logger, t.metrics)
-		if err != nil {
-			return nil, err
-		}
-		t.indexSets[id] = indexSet
+	var err error
+	baseIndexSet := t.baseUserIndexSet
+	if id == "" {
+		baseIndexSet = t.baseCommonIndexSet
+	}
 
+	// instantiate the index set, add it to the map and unlock the mutex before calling Init()
+	indexSet, err = NewIndexSet(t.name, id, filepath.Join(t.cacheLocation, id), baseIndexSet, t.boltDBIndexClient, t.logger, t.metrics)
+	if err != nil {
+		return nil, err
```

Comment:
Missing the Unlock() here. Which raises the question whether it would be a better idea to defer the `Unlock` and maybe factor a bit so that the expensive `indexSet.Init()` is not in the locked block.

```suggestion
        t.indexSetsMtx.Unlock()
		return nil, err
```

Reply:
ahh, sorry, missed it. It would be hard to use defer because with synchronous initialization it would be blocked anyway. I will take care of this issue.

Reply:
@chaudum can you please have a look at it now?